adapters/grpc/document_grpc.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 Nitric Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpc
+
+import (
+	"context"
+
+	pb "github.com/nitric-dev/membrane/interfaces/nitric/v1"
+	"github.com/nitric-dev/membrane/sdk"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// GRPC Interface for registered Nitric Document Plugin
+type DocumentServer struct {
+	pb.UnimplementedDocumentServer
+	// TODO: Support multiple plugin registrations
+	// Just need to settle on a way of addressing them on calls
+	documentPlugin sdk.DocumentService
+}
+
+func (s *DocumentServer) checkPluginRegistered() (bool, error) {
+	if s.documentPlugin == nil {
+		return false, status.Errorf(codes.Unimplemented, "Document plugin not registered")
+	}
+
+	return true, nil
+}
+
+func (s *DocumentServer) Set(ctx context.Context, req *pb.DocSetRequest) (*pb.DocSetResponse, error) {
+	if ok, err := s.checkPluginRegistered(); ok {
+		key := toSdkKey(req.Key)
+		subKey := toSdkKey(req.SubKey)
+
+		if err := s.documentPlugin.Set(*key, subKey, req.GetValue().AsMap()); err == nil {
+			return &pb.DocSetResponse{}, nil
+		} else {
+			// Case: Failed to create the key
+			// TODO: Translate from internal Document Service Error
+			return nil, err
+		}
+	} else {
+		// Case: Plugin was not registered
+		return nil, err
+	}
+}
+
+func (s *DocumentServer) Get(ctx context.Context, req *pb.DocGetRequest) (*pb.DocGetResponse, error) {
+	if ok, err := s.checkPluginRegistered(); ok {
+		key := toSdkKey(req.Key)
+		subKey := toSdkKey(req.SubKey)
+
+		if val, err := s.documentPlugin.Get(*key, subKey); err == nil {
```

Comment:
See comment on reference v. value, we can avoid having to de-reference the primary key in this case if we just take both by reference.

Reply:
sure